Develop

MakashevaPatina opened the pull request.

tim4054 committed:
ver1.0

tim4054 committed:
ver1.0

MakashevaPatina committed:
Creating a controller, service and DTO

russa89 committed:
v.1

tim4054 committed:
ver1.1
changed userId data type to String
renamed Recommendation to RecommendationDTO
injected repository

tim4054 committed:
ver1.2
changed getRecommendation() for repositories check methods data type BOOLEAN

russa89 committed:
v.2

russa89 committed:
v.2 renaming

tim4054 committed:
Merge remote-tracking branch 'origin/RecommendationRuleSetInterface' into develop

tim4054 committed:
Merge remote-tracking branch 'origin/repository' into develop

tim4054 committed:
contains all

MakashevaPatina committed:
Handling errors in the repository

tim4054 committed:
added service unit test.
controller test in progress

Fandom123023 committed:
test02

Fandom123023 committed:
test03

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,31 @@
+
+        List<RecommendationDTO> recommendations = ruleSets.stream()
+                .map(rule -> rule.getRecommendation(userId))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
```
можно faltMap(Optional::stream), но можно и так, в какой то мере даже понятнее и очевиднее

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/repository/RecommendationsRepository.java:
```diff
@@ -0,0 +1,105 @@
+                    Integer.class,
+                    user);
+            return result != null ? result : 0;
+        } catch (Exception e) {
+            return 0;
+        }
```
Не уверен что замазывать ошибки тут верно, мы так можем пропустить косяки в строении запроса или бд

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/repository/RecommendationsRepository.java:
```diff
@@ -0,0 +1,105 @@
+
+    public boolean checkRecommendationInvest500(String id) {
+
+        UUID uuid = UUID.fromString(id);
```
UUID может и не сойтись, я бы в контроллере принимал сразу UUID

tim4054 committed:
Dynamic

tim4054 committed:
Dynamic v2

MakashevaPatina committed:
adding recommendation logic

MakashevaPatina committed:
A
Merge branch 'develop' of https://github.com/MakashevaPatina/Bank_STAR into develop

MakashevaPatina committed:
added dynamic rules processing

MakashevaPatina committed:
dynamic rules processing

MakashevaPatina committed:
dynamic rules processing

MakashevaPatina committed:
RecommendationResponse

MakashevaPatina committed:
RecommendationResponse

MakashevaPatina committed:
RecommendationResponse

russa89 committed:
repository v.2.0

MakashevaPatina committed:
Merge branch 'repository' into develop

MakashevaPatina committed:
add productId and corrected logic

MakashevaPatina committed:
corrected model

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/configuration/RecommendationsDataSourceConfiguration.java:
```diff
@@ -26,4 +29,33 @@ public JdbcTemplate recommendationsJdbcTemplate(
         return new JdbcTemplate(dataSource);
     }
+
+    @Bean
+    public GroupedOpenApi allGroups() {
+        return GroupedOpenApi.builder().
+                group("Все контроллеры").
+                pathsToMatch("/**").
+                build();
+    }
+
+    @Bean
+    public GroupedOpenApi ruleGroup() {
+        return GroupedOpenApi.builder().
+                group("Контроллер правил").
+                pathsToMatch("/rule/**").
+                build();
+    }
```
можно в отдельный конфиг

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/repository/RecommendationsRepository.java:
```diff
@@ -0,0 +1,167 @@
+
+        UUID uuid = UUID.fromString(id);
+
+        if (getAmountOfTransactionTypeForProductTypeForUser(uuid, "DEBIT", "DEPOSIT") +
+                getAmountOfTransactionTypeForProductTypeForUser(uuid, "DEBIT", "WITHDRAW") > 0 &&
+
+                getAmountOfTransactionsForUserProductTypeExcluded(uuid, "INVEST") > 0 &&
+
+                getAmountOfTransactionTypeForProductTypeForUser(uuid, "SAVING", "DEPOSIT") > 1000) {
```
это бизнес логика, её не должно быть в репозитории, репозиторий только делает запросы

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/rules/dynamic/Condition.java:
```diff
@@ -0,0 +1,75 @@
+@Entity
+public class Condition {
+    @Id
+    @GeneratedValue
```
без стратегии специально?

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/service/DynamicRuleService.java:
```diff
@@ -0,0 +1,45 @@
+
+@Service
+public class DynamicRuleService {
+    Logger logger = LoggerFactory.getLogger(DynamicRule.class);
```
private final (static)

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/service/DynamicRuleService.java:
```diff
@@ -0,0 +1,45 @@
+            logger.info("Удаление правила c id:{}", id);
+            return ruleForDelete;
+        } catch (DynamicRuleNotFoundException e) {
+            logger.error("Правило с id:{} не найдено", id, e);
+            throw e;
```
на самом деле ошибка сама по себе лог, её не надо дублировать

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,147 @@
+    private final DynamicRulesRepository dynamicRulesRepository;
+    private final RecommendationsRepository recommendationsRepository;
+    private final JdbcTemplate jdbcTemplate;
+    Logger logger = LoggerFactory.getLogger(RecommendationService.class);
```
private final (static)

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,147 @@
+
+    private boolean handleUserOfQuery(String userId, List<String> arguments, boolean negate) {
+        String productType = arguments.get(0).toUpperCase();
+        String sql = "SELECT COUNT(*) FROM transactions WHERE user_id = ? AND product_id = (SELECT id FROM products WHERE type = ? LIMIT 1)";
```
А это наоборот ответственность репозитория

MakashevaPatina committed:
corrected service and repository

tim4054 committed:
telegramBot added

MakashevaPatina committed:
Recommendation rule triggering statistics

MakashevaPatina committed:
handling error when executing a command

tim4054 committed:
CommandRecommendation refactored

tim4054 committed:
CommandRecommendation refactored1

tim4054 committed:
CommandRecommendation refactored2

russa89 committed:
cacheController

Fandom123023 committed:
test05

MakashevaPatina committed:
Listener

MakashevaPatina committed:
Merge remote-tracking branch 'origin/develop01' into develop

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/CombinedApplication.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan(basePackages = {
+		"com.starbank.recommendation_service",
+		"com.starbank.telegram_bot"
+})
```
Если все пакеты находятся выше запускающего класса то ComponentScan внутри SpringBootApplication их по умолчанию захватит, это лишняя аннотация

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/configuration/CacheService.java:
```diff
@@ -0,0 +1,4 @@
+package com.starbank.recommendation_service.configuration;
+
+public class CacheService {
+}
```
удалить или доделать

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/telegramBot/listener/TelegramBotUpdatesListener.java:
```diff
@@ -0,0 +1,58 @@
+@Service
+public class TelegramBotUpdatesListener implements UpdatesListener {
+
+    private Logger logger = LoggerFactory.getLogger(TelegramBotUpdatesListener.class);
```
final и даже можно static

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/telegramBot/service/CommandService.java:
```diff
@@ -0,0 +1,50 @@
+
+    private Logger logger = LoggerFactory.getLogger(CommandRecommendation.class);
+
+    @Autowired
+    private TelegramBot telegramBot;
+
+
+    @Autowired
+    public CommandService(List<CommandHandler> handlers) {
+        handlers.forEach(handler -> commandHandlers.put(handler.getCommand(), handler));
+    }
```
тогда уж и бот можно было в конструктор, зачем отдельно
аннотация над конструктором не нужна

AleksandrZhukovJava commented on src/main/java/com/starbank/recommendation_service/telegramBot/service/CommandService.java:
```diff
@@ -0,0 +1,50 @@
+
+    public void handleCommand(Update update) {
+        String command = update.message().text();
+        CommandHandler handler = commandHandlers.get(command);
+        if (handler != null) {
+            handler.handle(update);
+        } else {
+            // Обработка неизвестной команды
+            Long chatID = update.message().chat().id();
+            SendMessage sendMessage = new SendMessage(chatID, "Unsupported command: " + command);
+            telegramBot.execute(sendMessage);
+            logger.info("Unsupported command: {}", command);
```
Можно сделать дефолтную команду и использовать getOrDefault

MakashevaPatina committed:
add diagram
